Log temperature

micahwoodard opened the pull request:
### Pull Request instructions:
- Please follow the [update protocol](https://github.com/AllenNeuralDynamics/aind-behavior-blog/wiki/Software-Update-Procedures)
- Answer the questions below in detail. Your responses will be emailed to experimenters. 
- If the experimenters must do anything new, provide detailed step by step instructions on the [wiki](https://github.com/AllenNeuralDynamics/aind-behavior-blog/wiki)
- If computer maintainers need to manually update anything, provide detailed step by step instructions
- Use markdown syntax in order for your comments to be rendered reliably in the email: "1." instead of "1)", use four spaces for indents.
- If you use the keyword "skip email" in the title, it will skip the email updates
- Merges from "develop" into "production_testing" should use the keyword "production merge" in the title for reliable indexing of updates
- Merges from "production_testing" into "main" should use the keyword "update main"
  
### Describe changes:
- Adds bonsai node to interface with the environment sensor on rigs. Also saves B_EnvironmentSensorTemperature, B_EnvironmentSensorHumidity, B_EnvironmentSensorPressure, and B_EnvironmentSensorTimestamp to the behavior.json. 
### What issues or discussions does this update address?
SIPE ticket [#4132](https://apps.powerapps.com/play/cb338490-e23d-48e6-9e6d-e600b8d17dd4?tenantId=32669cd6-737f-4b39-8bdd-d6951120d3fc&Hidenavbar=true&ItemID=4132)
### Describe the expected change in behavior from the perspective of the experimenter
- None
### Describe any manual update steps for task computers
- com port needs to be added to the Setting_box#.csv file. Example, for 1A EnvSen,COM3 needs to be added. 
### Was this update tested in 446/447?
- [x] 446/7
### Does this update impact downstream processing by adding new saved files, or changing their format? If so, have you documented changes?
- Adds B_EnvironmentSensorTemperature, B_EnvironmentSensorHumidity, B_EnvironmentSensorPressure, and B_EnvironmentSensorTimestamp to the behavior.json

micahwoodard commented:
@Ahad-Allen this will add B_EnvironmentSensorTemperature, B_EnvironmentSensorHumidity, B_EnvironmentSensorPressure, and B_EnvironmentSensorTimestamp to the behavior.json. I don't think that should break anything but wanted to double check

micahwoodard commented:
@XX-Yin here are the bonsai workflows added 
![image](https://github.com/user-attachments/assets/45a061cd-ad7a-4029-9a6b-f946a8c69624)
![image](https://github.com/user-attachments/assets/f8fde3b7-ed31-46c0-b036-c77f1fe7ec8c)

XX-Yin commented:
@micahwoodard what will happen if we don't have the environment sensor connected?

micahwoodard commented:
@XX-Yin 

> @XX-Yin here are the bonsai workflows added. I'm not sure if all rigs have environment sensors, and I'm unsure how to check in bonsai if a key exists or not. We could just skip the set up if there is no environment sensor configured

I'm not sure how to check if a key is in a dictionary in bonsai. Do you?

XX-Yin commented:
@micahwoodard 

Some general questions

1. what is the TrialEnd doing here?
2. When do we get the Temperature, Humidity, Pressure? How are they aligned with the harp timestamp? 
3. What is the frequency to get these parameters?

XX-Yin commented:
> @XX-Yin
> 
> > @XX-Yin here are the bonsai workflows added. I'm not sure if all rigs have environment sensors, and I'm unsure how to check in bonsai if a key exists or not. We could just skip the set up if there is no environment sensor configured
> 
> I'm not sure how to check if a key is in a dictionary in bonsai. Do you?

@micahwoodard You can use the CameraPreview node as an example

micahwoodard commented:
@XX-Yin 

> 1. what is the TrialEnd doing here?

It triggers the sensor to collect environment data after every trial end. Without it, the sensor would never be prompted to collect data after initialization. We could also use a timer or other events. TrialEnd made sense to me since other data is emited after each trial end

> 2. When do we get the Temperature, Humidity, Pressure? How are they aligned with the harp timestamp?

They are aligned with the TrialEnd signal. I will double check they are aligned with the harp timestamps

> 3. What is the frequency to get these parameters?

After every TrialEnd. We can change this if we want more or less.

micahwoodard commented:
@XX-Yin I've addressed concerns about the com port not being in the settings json with this check. 

![image](https://github.com/user-attachments/assets/7b47337c-389c-4da5-ad2f-7279890c4f9d)


This is the code for the timestamps. Looking at the values in the json that are saved, it seems like they are synced to the behavior timestamps. I'm not sure though. what do you think?
![image](https://github.com/user-attachments/assets/5210c79f-b188-4fa9-a4a4-10de6ae98ac3)
![image](https://github.com/user-attachments/assets/a9ccbe14-353e-4464-b9d4-17baa24abc02)

micahwoodard commented:
@nathan-mars that sounds good to me! Will update

@XX-Yin how are other data points aligned with the harp timestamps? I'm unsure of how to verify if the environment sensor timestamps are synced

Ahad-Allen commented:
I can update the capsules with the new fields and starting testing later this week

micahwoodard commented:
@XX-Yin so I should use the Behavior.TimestampedDigitalInputStat instead?

micahwoodard commented:
> @micahwoodard This is not precise. You should use code like this
> <img alt="image" width="1094" src="https://private-user-images.githubusercontent.com/109394934/454023813-42b2e92d-70b1-4e0e-800f-039955a3bbad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk4MTkwNDQsIm5iZiI6MTc0OTgxODc0NCwicGF0aCI6Ii8xMDkzOTQ5MzQvNDU0MDIzODEzLTQyYjJlOTJkLTcwYjEtNGUwZS04MDBmLTAzOTk1NWEzYmJhZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYxM1QxMjQ1NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNWU1NjAyY2VmNmFlNGU1Y2Q0ZWI4MjRhZWY5M2ZiNWI2NmE2OTcwMWVjMzg3YjU1ZWU0ZWMwZjBlNjk2Y2ZkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.y0VjqcFLhEcMltob6E5re1FivwZ6jIYrZETNFkM6Q0Y">

@XX-Yin would be nice to be able to merge this in for next weeks testing. I don't think I can use TimestampedDigitalInputStat because it seems to me this only happens when a lickspout is licked? What specific part of this workflow are you referencing and why is using the timestamps imprecise if that is what other data uses? It seems like the two are coming from similar sources. 
![image](https://github.com/user-attachments/assets/aff72261-c1e7-41c1-9006-bd2cb98aef89)

XX-Yin commented:
> > @micahwoodard This is not precise. You should use code like this
> > <img alt="image" width="1094" src="https://private-user-images.githubusercontent.com/109394934/454023813-42b2e92d-70b1-4e0e-800f-039955a3bbad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDk4MTkwNDQsIm5iZiI6MTc0OTgxODc0NCwicGF0aCI6Ii8xMDkzOTQ5MzQvNDU0MDIzODEzLTQyYjJlOTJkLTcwYjEtNGUwZS04MDBmLTAzOTk1NWEzYmJhZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYxM1QxMjQ1NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNWU1NjAyY2VmNmFlNGU1Y2Q0ZWI4MjRhZWY5M2ZiNWI2NmE2OTcwMWVjMzg3YjU1ZWU0ZWMwZjBlNjk2Y2ZkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.y0VjqcFLhEcMltob6E5re1FivwZ6jIYrZETNFkM6Q0Y">
> 
> @XX-Yin would be nice to be able to merge this in for next weeks testing. I don't think I can use TimestampedDigitalInputStat because it seems to me this only happens when a lickspout is licked? What specific part of this workflow are you referencing and why is using the timestamps imprecise if that is what other data uses? It seems like the two are coming from similar sources. ![image](https://private-user-images.githubusercontent.com/110491290/454876944-aff72261-c1e7-41c1-9006-bd2cb98aef89.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTAyMTMyMTksIm5iZiI6MTc1MDIxMjkxOSwicGF0aCI6Ii8xMTA0OTEyOTAvNDU0ODc2OTQ0LWFmZjcyMjYxLWMxZTctNDFjMS05MDA2LWJkMmNiOThhZWY4OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNjE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDYxOFQwMjE1MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lN2NlZTRhNTlkMjRlMmVkOWQxNzU0OWQxMDRkODI2NmZkY2ZhZjkxYTkwOThmZTE5ZDhlN2EwY2M3NGMxODRjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.zIgm6hU6ClDoMekecomfadoLnM5X0IrTN1gkndoxWQM)

@micahwoodard To directly pair the environment sensor event with a timestamp is not precise as they are not synchronized. 

1. Is the environment sensor a HARP device? If so, could you add it to the logging node?

2. It would be preferable to collect measurements at a fixed frequency rather than at the end of each trial, since trial durations are varied. That said, perhaps this doesn’t matter much if variables like temperature and humidity remain relatively stable. Do you have any data showing the stability of these measurements?

nathan-mars commented:
>@micahwoodard It’s not accurate to directly pair environment sensor events with timestamps, as they are not synchronized.
> 1. Is the environment sensor a HARP device? If so, could you add it to the logging node?

It is a Harp device with a 3.5mm sync line. I don't understand the comment that env sensor events aren't synchronized with timestamps, but I'll leave that to others to sort out. 

> 2. It would be preferable to collect measurements at a fixed frequency rather than at the end of each trial, since trial durations are varied. That said, perhaps this doesn’t matter much if variables like temperature and humidity remain relatively stable. Do you have any data showing the stability of these measurements?

Multiple measurements should be made, not just the start/end temperature. Ideally once a second or something convenient to implement. One of the reasons for adding this to the workflow is so that we can test how stable temperature/etc are.

bruno-f-cruz commented:
I already talked to @micahwoodard about this. This is a harp-device as such it can be logged like ANY harp device (https://allenneuraldynamics.github.io/Bonsai.AllenNeuralDynamics/articles/core-logging.html#with-metadata). Is there a reason to not follow the standard?

The device is indeed a flully fledged harp device, it collects data at a periodic rate (https://allenneuraldynamics.github.io/Bonsai.AllenNeuralDynamics/articles/environment-sensor.html#parsing-sensor-data) and is synchronized with all other harp devices in the rig provided a synch cable is provided.

micahwoodard commented:
Great thanks so much for everyone's input!

> 1. If so, could you add it to the logging node?

> I already talked to @micahwoodard about this. This is a harp-device as such it can be logged like ANY harp device (https://allenneuraldynamics.github.io/Bonsai.AllenNeuralDynamics/articles/core-logging.html#with-metadata). Is there a reason to not follow the standard?

@XX-Yin where is the logging node? Is it the same as what @bruno-f-cruz is referencing? Do we not need to save values in the behavior json if we save them in the log?

> The device is indeed a flully fledged harp device, it collects data at a periodic rate (https://allenneuraldynamics.github.io/Bonsai.AllenNeuralDynamics/articles/environment-sensor.html#parsing-sensor-data) and is synchronized with all other harp devices in the rig provided a synch cable is provided.

@bruno-f-cruz does that mean that the timestamps coming from the sensor are already synchronized so I can just use those directly? I do not have a lot of experience with harp devices or bonsai or hardware on the behavior rigs

micahwoodard commented:
@XX-Yin anything else I need to do to get this merged in?

XX-Yin commented:
> @XX-Yin anything else I need to do to get this merged in?

@micahwoodard  Have you tested the updated version?

XX-Yin commented on src/foraging_gui/MyFunctions.py:
```diff
@@ -3033,6 +3037,19 @@ def _get_irregular_timestamp(self, Channel2):
                     self.B_AutoRightWaterStartTime, Rec[1][1][0]
                 )
+            elif Rec[0].address == "/EnvironmentSensorTemperature":
+                value = Rec[1][1][0] if type(Rec[1][1][0]) != float else round(Rec[1][1][0], 1)
```
is it necessary to round?

micahwoodard replied:
In earlier comments, it was decided to round 

from @alexpiet 
> @micahwoodard @nathan-mars The raw data from the sensors look like they have 16 digits after the decimal point. What is the actual precision of these sensors? I'm not sure if its best practice to just log everything and deal with it later, or round to the actual precision now.

from @nathan-mars 
> The datasheet says ±0.5ºC and ±1.5% RH accuracy ([BME688](https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bme688-ds000.pdf)). (Side note - we haven't calibrated the relative humidity, so I would not trust that value yet).
> Keeping 1 past decimal makes sense to me (23.45678 recorded as 23.5)

micahwoodard commented:
had to change layout file since I was getting this error with old one. Double checked on rig that layout still looks good 
![image](https://github.com/user-attachments/assets/4c67bf22-03af-4866-be8c-b96d2bd63705)
